### DISCUSSION TITLE
code completion setup example with Company mode please

### DISCUSSION THREAD

**Discussion by suntong** [emacs-lsp/lsp-mode#2937] [Q&A] [answered]
I've done lots of reading on code completion setup using `lsp-mode` and `company` mode, but haven't found a working one yet, as most that I read are still using `company-lsp` to do the integration, which is now deprecated and no longer works with lsp-mode.

So, appreciate somebody giving me a working setup example as I don't know elisp except copying from working code.
Even a point to a github repo can do. Better if it is for Go.

Thx

**Comment by yyoncho** [comment, marked as answer]
There is no additional setup, lsp-mode is doing everything automatically, just make sure that you have the latest version of lsp-mode and company-mode.

**Reply by suntong** [reply]
Interesting. I do have the latest version of lsp-mode and company-mode, installed just few days ago, 

So you're saying a setup as simple as this:

```el
(use-package lsp-mode
  :ensure t)
(use-package company
  :ensure t)
```

will have code completion out of the box?

How does it work then?

For `fmt.Println()`, I want to type `f` or `fm` and get prompted for `fmt`, and when I've typed `fmt.P`, I want to see the pop up of a list of functions to choose from. 

But I'm not seeing any of them. 
What I'm missing? thx

**Reply by kiennq** [reply]
Did you enable `lsp` inside the buffer?
Also, for typing `f` and `fm` and have the completion shown, you will need to customize `company-mode` to have `(setq company-minimum-prefix-length 1)`.
That's `company-mode` setting and not really related to `lsp-mode` though.

**Reply by suntong** [reply]
Still not able to make it work for me.

Here is almost [all of my `init.el`](https://github.com/suntong/emacs.d/blob/master/init.mini.el), apart from `load-path` setting:

```el
;; Company mode is a standard completion package that works well with lsp-mode.
(use-package company
  :ensure t
  :config
  ;; Optionally enable completion-as-you-type behavior.
  ;; don't add any dely before trying to complete thing being typed
  ;; the call/response to gopls is asynchronous so this should have little
  ;; to no affect on edit latency
  (setq company-idle-delay 0.1)
  ;; start completing after a single character instead of 3
  (setq company-minimum-prefix-length 1)
  ;; align fields in completions
  (setq company-tooltip-align-annotations t)
  )
)

;; lsp-mode will detect company-mode automatically and will use it if its installed
(use-package lsp-mode
  :ensure t
  :commands (lsp lsp-deferred)
  :hook (go-mode . lsp-deferred)
  :config
  ;; The CAPF back-end provides a bridge to the standard completion-at-point-functions facility, and thus works with any major mode that defines a proper completion function.
  (setq lsp-completion-provider :capf)
  (add-hook 'go-mode-hook #'lsp-go-install-save-hooks))
```

But I don't get anything that I described above:

![image](https://user-images.githubusercontent.com/422244/122681682-909d7c00-d1c3-11eb-9091-6d73608a56b1.png)

![image](https://user-images.githubusercontent.com/422244/122681797-1c170d00-d1c4-11eb-99c7-17cd137433e8.png)

**Reply by suntong** [reply]
I should have lsp enabled inside the buffer, right?

![image](https://user-images.githubusercontent.com/422244/122681887-7a43f000-d1c4-11eb-80d9-56d11cbecf8e.png)

Yeah, I think so:

![image](https://user-images.githubusercontent.com/422244/122681947-c98a2080-d1c4-11eb-97a6-ac9000efa871.png)

**Reply by yyoncho** [reply]
I don't think that gopls will work in a buffer in tmp folder. Gopls requires the projects to be under GO_SRC(or something like that).

**Reply by suntong** [reply]
OK, I've move it to `$GOPATH/src/helloworld`, and I've added the following to my `init.el`:

```el
(use-package go-mode
  :ensure t
  :defer t
  :hook (go-mode . (lambda ()
                     (require 'lsp-go)
                     (require 'dap-go)
                     (lsp-deferred))))
```

But this is what I'm getting (gopls error buffer is empty):

![image](https://user-images.githubusercontent.com/422244/122683264-19b8b100-d1cc-11eb-89fa-85091b6c9f1d.png)

Still no code completion 😢😢

**Reply by yyoncho** [reply]
Do you have `import "fmt"`? What does company-diag returns? Have you checked the logs if there is a completion request sent to the server(instructions on how to do that in the bug report template)?

**Reply by suntong** [reply]
> Do you have import "fmt"?

Yes, I do. See the first screenshot. It's the same file for all screenshots.

`emacs -q -l /tmp/lsp-start-plain.el` looks good, `company-diag` returns normal ... 

Wait, it is working now....

Hmm... maybe my `(use-package company` have an extra "`)`" at the end, but my `emacs --debug-init` somehow didn't detected it before...

But anyway, it is working now. 
Sorry for such a silly mistake.

Thanks everyone for your help.

**Reply by suntong** [reply]
> OK, I've move it to `$GOPATH/src`...

Found this was the main reason everything is working for me now. 

Previously all my code are not under `$GOPATH`
